File: common/logisticspipes/commands/abstracts/SubCommandHandler.java
package logisticspipes.commands.abstracts;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import net.minecraft.command.ICommandSender;

import logisticspipes.commands.chathelper.MorePageDisplay;
import logisticspipes.commands.exception.CommandNotFoundException;
import logisticspipes.commands.exception.DublicatedCommandException;
import logisticspipes.commands.exception.MissingArgumentException;
import logisticspipes.commands.exception.PermissionDeniedException;
import logisticspipes.utils.string.ChatColor;

public abstract class SubCommandHandler implements ICommandHandler {

	public SubCommandHandler() {
		registerSubCommands();
	}

	private final List<ICommandHandler> subCommands = new ArrayList<>();

	public abstract void registerSubCommands();

	protected final void registerSubCommand(ICommandHandler newHandler) {
		List<String> newCommands = Arrays.asList(newHandler.getNames());
		for (ICommandHandler handler : subCommands) {
			for (String command : handler.getNames()) {
				if (newCommands.contains(command)) {
					throw new DublicatedCommandException();
				}
			}
		}
		subCommands.add(newHandler);
	}

	public final void displayHelp(ICommandSender sender) {
		MorePageDisplay display = new MorePageDisplay(new String[] { "|< Help - " + getNames()[0] + " - Page: %/$ >|" }, sender);
		for (ICommandHandler command : subCommands) {
			if (!command.getDescription()[0].startsWith("#")) {
				boolean first = true;
				String prefix = (command instanceof SubCommandHandler ? ChatColor.BLUE : ChatColor.YELLOW) + command.getNames()[0] + ChatColor.RESET + ": ";
				for (int d = 0; d < command.getDescription().length; d++) {
					display.append(prefix + command.getDescription()[d], !first);
					prefix = "    ";
					first = false;
				}
				if (command instanceof SubCommandHandler) {
					display.append("      " + ChatColor.GRAY + "- add " + ChatColor.YELLOW + "help" + ChatColor.GRAY + " to see the subcommands", true);
				}
				first = true;
				if (command.getNames().length > 1) {
					for (int i = 1; i < command.getNames().length; i++) {
						display.append((first ? "  alias: - " : "         - ") + ChatColor.GOLD + command.getNames()[i] + "", true);
						first = false;
					}
				}
				display.append("", true);
			}
		}
		display.display(sender);
	}

	@Override
	public final void executeCommand(ICommandSender sender, String[] args) {
		if (args.length < 1) {
			throw new MissingArgumentException();
		}
		String command = args[0];
		if (command.equalsIgnoreCase("help")) {
			displayHelp(sender);
			return;
		} else {
			for (ICommandHandler handler : subCommands) {
				if (Arrays.asList(handler.getNames()).contains(command)) {
					if (!handler.isCommandUsableBy(sender)) {
						throw new PermissionDeniedException();
					}
					String[] newArgs = Arrays.copyOfRange(args, 1, args.length);
					handler.executeCommand(sender, newArgs);
					return;
				}
			}
			throw new CommandNotFoundException();
		}
	}
}

File: common/logisticspipes/commands/abstracts/SubCommandHandler.kt
package logisticspipes.commands.abstracts;

import net.minecraft.command.ICommandSender;

import logisticspipes.commands.chathelper.MorePageDisplay;
import logisticspipes.commands.exception.CommandNotFoundException;
import logisticspipes.commands.exception.DublicatedCommandException;
import logisticspipes.commands.exception.MissingArgumentException;
import logisticspipes.commands.exception.PermissionDeniedException;
import logisticspipes.utils.string.ChatColor;

abstract class SubCommandHandler : ICommandHandler {

	private val subCommands: MutableList<ICommandHandler> = mutableListOf()

	protected fun registerSubCommand(newHandler: ICommandHandler) {
		val newCommands = newHandler.getNames().toList()
		for (handler in subCommands) {
			for (command in handler.getNames()) {
				if (command in newCommands) {
					throw DublicatedCommandException()
				}
			}
		}
		subCommands.add(newHandler)
	}

	private fun displayHelp(sender: ICommandSender) {
		val display = MorePageDisplay(arrayListOf("|< Help - " + getNames()[0] + " - Page: %/$ >|"), sender)
		for (handler in subCommands) {
			if (handler.getDescription()[0].startsWith("#")) continue

			val color = if (handler is SubCommandHandler) ChatColor.BLUE else ChatColor.YELLOW;
			var prefix = "$color${handler.getNames()[0]}${ChatColor.RESET}: "

			for (d in handler.getDescription().indices) {
				display.append(prefix + handler.getDescription()[d], d > 0)
				prefix = "    "
			}
			if (handler is SubCommandHandler) {
				display.append("$prefix${ChatColor.GRAY}- add ${ChatColor.YELLOW}help${ChatColor.GRAY} to see the subcommands", true)
			}

			if (handler.getNames().size > 1) {
				for (i in 1..handler.getNames().indices.last) {
					display.append((if (i == 1) "  alias: - " else "         - ") + "${ChatColor.GOLD}${handler.getNames()[i]}", true)
				}
			}
			display.append("", true)
		}
		display.display(sender)
	}

	final override fun executeCommand(sender: ICommandSender, args: Array<String>) {
		if (args.isEmpty()) {
			throw MissingArgumentException()
		}
		val command = args[0]
		if (command.equals("help", true)) {
			displayHelp(sender)
			return
		}
		for (handler: ICommandHandler in subCommands) {
			if (command in handler.getNames()) {
				if (!handler.isCommandUsableBy(sender)) {
					throw PermissionDeniedException()
				}
				handler.executeCommand(sender, args.copyOfRange(1, args.size))
				return
			}
		}
		throw CommandNotFoundException()
	}
}
